docs/docs/typelevel.md

```diff
@@ -0,0 +1,449 @@
+of creating a `TreeSet[T]` or a `HashSet[T]` depending on whether `T` has an `Ordering` or not. We can create a set of implicit definitions like this:
+```scala
+trait SetFor[T, S <: Set[T]]
+class LowPriority {
+  implicit def hashSetFor[T]: SetFor[T, HashSet[T]] = ...
+}
+object SetsFor extends LowPriority {
+  implicit def treeSetFor[T: Ordering]: SetFor[T, TreeSet[T]] = ...
+}
+```
+Clearly, this is not pretty. Besides all the usual indirection of implicit search,
+we face the problem of rule prioritization where
+we have to ensure that `treeSetFor` takes priority over `hashSetFor` if the element type has an ordering. This is solved
+(clumsily) by putting `hashSetFor` in a superclass `LowPriority` of the object `SetsFor` where `treeSetFor` is defined.
+Maybe the boilerplate would still be acceptable if the crufty code could be contained. However, this is not the case. Every user of
+the abstraction has to be parameterized itself with a `SetFor` implicit. Considering the simple task _"I want a `TreeSet[T]` if
+`T` has an ordering and a `HashSet[T]` otherwise"_, this seems like a lot of ceremony.
+
+There are some proposals to improve the situation in specific areas, for instance by allowing more elaborate schemes to specify priorities. But they all keep the viral nature of implicit search programs based on logic programming.
+
+By contrast, the new `implicit match` construct makes implicit search available in a functional context. To solve
+the problem of creating the right set, one would use it as follows:
+```scala
+transparent def setFor[T]: Set[T] = implicit match {
+  case Ordering[T] => new TreeSet[T]
+  case _           => new HashSet[T]
+}
+```
+An implicit match uses the `implicit` keyword in the place of the scrutinee. Its patterns are types.
+Patterns are tried in sequence. The first case with a pattern for which an implicit can be summoned is chosen.
+
+Implicit matches can only occur as toplevel match expressions of transparent methods. This ensures that
+all implicit searches are done at compile-time.
+
+There may be several patterns in a case of an implicit match, separated by commas. Patterns may bind type variables.
+For instance the case
+```scala
+  case Convertible[T, u], Printable[`u`] => ...
+```
+matches any type `T` that is `Convertible` to another type `u` such that `u` is `Printable`.
+
+## New Syntax for Type Variables in Patterns
+
+This last example exhibits some awkwardness in the way type variables are currently handled in patterns. To _bind_ a type variable,
+it needs to start with a lower case letter, then to refer to the bound variable itself, the variable has to be
+put in backticks. This is doable, but feels a bit unnatural. To improve on the syntax, we
+allow an alternative syntax that prefixes type variables by `type`:
+```scala
+  case Convertible[T, type U], Printable[U] => ...
```

Comment:
Will an analogous ability be providing for matching on values where the same value is bound in two places? (e.g. tuple2 match { case (x, x) => })

Reply:
Also, matching on two patterns is something that is useful in value pattern matches too. It would be nice to have support for that. Currently one has to use something like

```scala
object & {
  def unapply[A](a: A): Option[(A, A)] = Some((a, a))
}

thing match { case FirstPattern(a) & SecondPattern(b, c) => }
```

It would be nice if (1) there would be something like that out of the box, and (2) there would be symmetry between however it is done with type patterns and with value patterns.

Reply:
That's definitely worth some thought!

Reply:
@odersky Why not generalize `@` so it can take a pattern on the left-hand side, and not just an identifier?
Then we can use it both for values:
```scala
thing match { case FirstPattern(a) @ SecondPattern(b, c) => }
```
And for types:
```scala
case Convertible[T, type U] @ Printable[U] => ...
```
Adding to the pattern-matching wish-list, it would be super useful and powerful if we could have repeated occurrences of pattern variables, with these properties:

 * a pattern variable occurring several times in an 'AND' pattern gets the type of the first occurrence, and all extracted occurrences are checked for equality on extraction. For example `case (Foo(x),Bar(x))` is equivalent to `case (Foo(x),Bar(x0)) if x == x0`.

 * a pattern variable occurring several times in an 'OR' pattern gets the least upper bound type of the occurrences. For example `case Foo(x) | Bar(x)` extracts `x: A | B` if `Foo.unapply` extracts an `A` and `Bar.unapply` extracts a `B`.

 * if a pattern variable typed `T` does not occur in all branches of an 'OR' pattern, it gets wrapped in an option as `Option[T]` --- or alternatively, to make it more consistent, it gets type `T | Unit` or even `T | null` if/when Scala supports`null`-checking.